extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -303,70 +252,128 @@ export class AmpA4A extends AMP.BaseElement {
+        // This block returns the ad URL, if one is available.
+        .then(() => {
+          checkStillCurrent(promiseId);
+          return this.getAdUrl();
+        })
+        // This block returns the (possibly empty) response to the XHR request.
+        .then(adUrl => {
+          checkStillCurrent(promiseId);
+          this.adUrl_ = adUrl;
+          return adUrl && this.sendXhrRequest_(adUrl);
+        })
+        // The following block returns either the response (as a {bytes, headers}
+        // object), or null if no response is available / response is empty.
+        .then(fetchResponse => {
+          checkStillCurrent(promiseId);
+          if (!fetchResponse || !fetchResponse.arrayBuffer) {
+            return null;
+          }
+          // Note: Resolving a .then inside a .then because we need to capture
+          // two fields of fetchResponse, one of which is, itself, a promise,
+          // and one of which isn't.  If we just return
+          // fetchResponse.arrayBuffer(), the next step in the chain will
+          // resolve it to a concrete value, but we'll lose track of
+          // fetchResponse.headers.
+          return fetchResponse.arrayBuffer().then(bytes => {
+            return {
+              bytes,
+              headers: fetchResponse.headers,
+            };
+          });
+        })
+        // This block returns the ad creative and signature, if available; null
+        // otherwise.
+        .then(responseParts => {
+          checkStillCurrent(promiseId);
+          return responseParts && this.extractCreativeAndSignature(
+              responseParts.bytes, responseParts.headers);
+        })
+        // This block returns the ad creative if it exists and validates as AMP;
+        // null otherwise.
+        .then(creativeParts => {
+          checkStillCurrent(promiseId);
+          if (!creativeParts || !creativeParts.signature) {
+            return null;
+          }
+
+          // Alias for validCreativePromise's resolve function. This will be called
+          // if any of the fetched keys successfully validate the creative.
+          let resolveValidation;
+          // Promise that will resolve upon any successful validation.
+          const validCreativePromise = new Promise(resolve => {
+            // @param {!function(?ArrayBuffer)} resolve
+            resolveValidation = resolve;
+          });
+
+          // The following block defines an Array of Promises of ArrayBuffers,
+          // none of which will ever resolve. In the first instance of a
+          // successful creative verification, resolveValidation will be
+          // called with the creative as the argument.
+          const signatureVerificationPromises = this.keyInfoSetPromises_.map(
+              keyInfoSetPromise => {
+                // @param {!Promise<!Array<!Promise<?PublicKeyInfoDef>>>}
+                // keyInfoSetPromise
+                // @return {!Promise}
```

Comment:
These aren't hinted correctly. If you want to typecast the return value, you need:

```js
return /** @type {!Promise} */ X
```

In general though, these aren't needed.